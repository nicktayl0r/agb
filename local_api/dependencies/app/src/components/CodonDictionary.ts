export const CODON_DICTIONARY: { [index: string]: string } = {
  UUU: "Phenylalanine",
  UUC: "Phenylalanine",
  UUA: "Leucine",
  UUG: "Leucine",
  CUU: "Leucine",
  CUC: "Leucine",
  CUA: "Leucine",
  CUG: "Leucine",
  AUU: "Isoleucine",
  AUC: "Isoleucine",
  AUA: "Isoleucine",
  AUG: "Methionine",
  GUU: "Valine",
  GUC: "Valine",
  GUA: "Valine",
  GUG: "Valine",
  UCU: "Serine",
  UCC: "Serine",
  UCA: "Serine",
  UCG: "Serine",
  CCU: "Proline",
  CCC: "Proline",
  CCA: "Proline",
  CCG: "Proline",
  ACU: "Threonine",
  ACC: "Threonine",
  ACA: "Threonine",
  ACG: "Threonine",
  GCU: "Alanine",
  GCC: "Alanine",
  GCA: "Alanine",
  GCG: "Alanine",
  UAU: "Tyrosine",
  UAC: "Tyrosine",
  UAA: "Stop",
  UAG: "Stop",
  CAU: "Histidine",
  CAC: "Histidine",
  CAA: "Glutamine",
  CAG: "Glutamine",
  AAU: "Asparagine",
  AAC: "Asparagine",
  AAA: "Lysine",
  AAG: "Lysine",
  GAU: "Aspartic acid",
  GAC: "Aspartic acid",
  GAA: "Glutamic acid",
  GAG: "Glutamic acid",
  UGU: "Cysteine",
  UGC: "Cysteine",
  UGA: "Stop_UGA",
  UGG: "Tryptophan",
  CGU: "Arginine",
  CGC: "Arginine",
  CGA: "Arginine",
  CGG: "Arginine",
  AGU: "Serine",
  AGC: "Serine",
  AGA: "Arginine",
  AGG: "Arginine",
  GGU: "Glycine",
  GGC: "Glycine",
  GGA: "Glycine",
  GGG: "Glycine"
};
