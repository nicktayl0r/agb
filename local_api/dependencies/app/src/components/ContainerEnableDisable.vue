<template>
  <fieldset :id="id" v-show="visibleStore" :disabled="disabled" :class="[ 'no-border', AddRemoveClasses]" ><slot/></fieldset>
</template>

<script lang="ts">
import { Component } from "vue-property-decorator";
import Widget from "./Widget";
import WidgetEnabled from "../mixins/WidgetEnabled";

@Component({ mixins: [WidgetEnabled] })
export default class ContainerEnableDisable extends Widget {}
</script>
