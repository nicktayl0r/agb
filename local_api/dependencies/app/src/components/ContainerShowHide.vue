<template>
  <div :id="id" v-show="visibleStore" :class="['container', AddRemoveClasses]"><slot/></div>
</template>

<style>
.container {
  width: 100%;
}
</style>

<script lang="ts">
import { Component } from "vue-property-decorator";
import Widget from "./Widget";

@Component
export default class ContainerShowHide extends Widget {}
</script>
