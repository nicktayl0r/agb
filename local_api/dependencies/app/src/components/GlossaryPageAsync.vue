<template>
  <component
    :is="page"
    :projectIsReady="projectLoaded"
    :pageID="glossaryPageID"
    class="glossaryPage"
  />
</template>

<script lang="ts">
import { AxiosResponse } from "axios";
import { AxiosRetryInstance } from "@/helpers/requestHelpers";
import { Component, Prop } from "vue-property-decorator";


import Page from "./Page";
import { logMessage } from "@/helpers/debugHelpers";
import { tryCatch } from "@/helpers/projectHelpers";

import { getEnvPageURL } from "@/helpers/envHelpers";
import { readProjectDataExists } from "@/store/modules/appData";
import { readModulesState } from "@/store/modules/userData";

@Component
export default class GlossaryPageAsync extends Page {
  page = Page.extend({ template: "<div>loading...</div>" });

  @Prop({ default: false })
  projectIsReady: boolean;

  @Prop()
  pageID: string;

  projectNowReady = false;

  get projectLoaded() {
    logMessage("glossary projectLoaded computed property detects change");
    const storeProject = readProjectDataExists(this.$store);
    if (!this.projectNowReady && storeProject) {
      this.projectNowReady = true;
      const pageID = this.glossaryPageID; // this will call loadPageIfValid
    }
    return storeProject;
  }

  get glossaryPageID(): string {
    const modulesState = readModulesState(this.$store);
    const pageID = modulesState.currentGlossaryPageID;
    this.loadPageIfValid(pageID);
    return pageID;
  }

  loadPageIfValid(pageID: string) {
    if (pageID) {
      this.getPageAsync(pageID);
    }
  }

  getPageAsync(pageID: string) {
    logMessage("attempting to async load glossary page", pageID);
    AxiosRetryInstance.get(getEnvPageURL(pageID), {
      responseType: "text"
    })
      .then(response => tryCatch(() => this.setPage(pageID, response)))
      .catch(error => tryCatch(() => {
        this.page = Page.extend({
          template: `<div>Couldn't load page ${pageID} because ${error}</div>`
        });
      }));
  }

  setPage(pageID: string, response: AxiosResponse) {
    const wrappedTemplate = `<div>${response.data}</div>`; // Vue wants templates to have a single parent element
    this.page = Page.extend({ template: wrappedTemplate });
  }
}
</script>

<style scoped>
.glossaryPage {
  height: 100%;
}
</style>
