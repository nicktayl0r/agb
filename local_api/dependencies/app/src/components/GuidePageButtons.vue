<template>
  <div v-show="guideOpen" :data-currentPage="guideCurrentPage">
    <widget-button id="guidePreviousPage" ref="guidePreviousPage" class="buttonBack" v-show="showPrevButton" :clickEffects="{conditionList:[ { effectsPass:[{effectData:{effectType: 'previousPage'}}] } ]}" />
    <widget-button id="guideNextPage" ref="guideNextPage" class="buttonForward" v-show="showNextButton" :class="{ buttonBgImgBlinkBlue: blinkNextPage() }" :clickEffects="{conditionList:[ { effectsPass:[{effectData:{effectType: 'nextPage'}}] } ]}" />
  </div>
</template>

<script lang="ts">
import { Component, Watch, Prop } from "vue-property-decorator";
import Vue, { VueConstructor } from "vue";
import { Maybe } from "true-myth";

import { readModulesState } from "../store/modules/userData";
import {
  getModuleFirstPageID,
  getModuleLastPageID
} from "../helpers/projectHelpers";
import Widget from "./Widget";
import WidgetButton from "./WidgetButton.vue";
import { logMessage } from "@/helpers/debugHelpers";
import {
  getNextPageID,
  isPageUnlocked,
  getNextPageUserState
} from "../helpers/userHelpers";
import { useTeacherView } from "@/helpers/widgetHelpers";

@Component({
  components: {
    WidgetButton
  }
})
export default class GuidePageButtons extends Vue {
  teacherView = false;

  guideCurrentPage = "";

  // for most widgets on pages we set their pageID on created
  // and we don't want them to navigate if the current page isn't their page
  // (if an effect is delayed, and the user navigates somewhere else)
  // since these buttons are special and exists outside of a page,
  // let's update their pageID whenever the current guide page changes in the store
  getGuideCurrentPage() {
    logMessage("GuidePageButtons.getGuideCurrentPage()");
    const modulesState = readModulesState(this.$store);
    const prevPageButton = this.$refs.guidePreviousPage as Widget;
    const nextPageButton = this.$refs.guideNextPage as Widget;

    this.guideCurrentPage = modulesState.guideState.currentPageID;
    if (prevPageButton) prevPageButton.pageID = this.guideCurrentPage;
    else {
      logMessage(
        "GuidePageButtons.prevPageButton not available cannot set guideCurrentPage"
      );
    }
    if (nextPageButton) nextPageButton.pageID = this.guideCurrentPage;
    else {
      logMessage(
        "GuidePageButtons.nextPageButton not available cannot set guideCurrentPage"
      );
    }
  }

  created() {
    this.teacherView = useTeacherView();
  }

  mounted() {
    this.getGuideCurrentPage();
  }

  get guideOpen() {
    this.getGuideCurrentPage();
    const modulesState = readModulesState(this.$store);
    return modulesState.moduleOpen === "guide";
  }

  get showPrevButton() {
    this.getGuideCurrentPage();
    const maybePageID = getModuleFirstPageID("guide");
    if (maybePageID.isJust()) {
      const pageID = Maybe.unsafelyUnwrap(maybePageID);
      logMessage(pageID);

      const modulesState = readModulesState(this.$store);
      return modulesState.guideState.currentPageID !== pageID;
    }
    return true;
  }

  get showNextButton() {
    this.getGuideCurrentPage();
    const maybeLastPageID = getModuleLastPageID("guide");
    if (maybeLastPageID.isJust()) {
      const lastPageID = Maybe.unsafelyUnwrap(maybeLastPageID);
      const modulesState = readModulesState(this.$store);
      if (modulesState.guideState.currentPageID === lastPageID) return false;
    }

    if (this.teacherView) return true;

    const maybeNextPageID = getNextPageID();
    if (maybeNextPageID.isJust()) {
      const nextPageID = Maybe.unsafelyUnwrap(maybeNextPageID);
      return isPageUnlocked(nextPageID);
    }

    return true;
  }

  blinkNextPage(): boolean {
    const maybeNextPageID = getNextPageID();
    if (maybeNextPageID.isJust()) {
      const nextPageID = Maybe.unsafelyUnwrap(maybeNextPageID);
      const unlocked = isPageUnlocked(nextPageID);

      let visitedPageBefore = false;
      const maybePageUserState = getNextPageUserState();
      if (maybePageUserState.isJust()) {
        const pageState = Maybe.unsafelyUnwrap(maybePageUserState);
        if (pageState.visited != undefined) visitedPageBefore = pageState.visited;
      }
      const blink = !visitedPageBefore && unlocked;
      return blink;
    }
    return false;
  }
}
</script>

<style>
#guidePreviousPage {
  position: absolute;
  bottom: 1.137vh;
  left: 1.725vh;
  pointer-events: all;
}
#guidePreviousPage .bgImage,
#guideNextPage .bgImage {
  width: 6vh;
  height: 6vh;
}

#guideNextPage {
  position: absolute;
  bottom: 1.137vh;
  right: 1.725vh;
  pointer-events: all;
}
</style>