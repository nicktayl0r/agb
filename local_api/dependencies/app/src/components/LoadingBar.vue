<template>
  <div class="loadingBarParent">
    <div class="loadingBar">
      <div class="loadingBarSpinner"></div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop } from "vue-property-decorator";
import Vue from "vue";
const ProgressBar = require("progressbar.js");

@Component
export default class LoadingBar extends Vue {
  progressBar: any;

  mounted() {
    const barElement = this.$el.querySelectorAll(".loadingBar")[0];
    this.progressBar = new ProgressBar.Line(barElement, {
      strokeWidth: 5,
      color: "#337BB4",
      from: { color: "#337BB4", width: 2 },
      to: { color: "#F28446", width: 5 },
      trailWidth: 0.8,
      step(state: any, line: any) {
        line.path.setAttribute("stroke", state.color);
        line.path.setAttribute("stroke-width", state.width);
        if (line.text) line.text.style.color = state.color;
        const value = Math.round(line.value() * 100);
        if (value === 0) {
          line.setText("Loading");
        } else {
          line.setText(`${value}%`);
        }
      },
      text: {
        className: "loadingBarText",
        value: "Loading",
        style: {
          padding: "10px"
        }
      },
      duration: 10
    });
  }

  beforeDestroy() {
    this.progressBar.destroy();
  }
}
</script>


<style>
.loadingBarParent {
  height: 100%;
  width: 100%;
  position: inherit;
  display: flex;
  background-image: url("../assets/svg/circleBg.svg");
}
.loadingBar {
  max-width: 50%;
  align-self: center;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
}
.loadingBarText {
  text-align: center;
}
.loadingBarSpinner {
  align-self: center;
  margin: 0 auto;
  width: 18.15vh;
  height: 18.15vh;
  background-image: url("../assets/svg/dnaSpinner.svg");
  background-repeat: no-repeat;
  background-size: contain;
}
</style>