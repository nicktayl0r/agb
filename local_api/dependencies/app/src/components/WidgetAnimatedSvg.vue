// Component Wrapper for lottie-web
/* Documentation: https://github.com/airbnb/lottie-web
Events
  onComplete
  onLoopComplete
  onEnterFrame
  onSegmentStart

you can also use addEventListener with the following events:
  complete
  loopComplete
  enterFrame
  segmentStart
  config_ready (when initial config is done)
  data_ready (when all parts of the animation have been loaded)
  loaded_images (when all image loads have either succeeded or errored)
  DOMLoaded (when elements have been added to the DOM)
  destroy
*/

<template>
	<div :id="id" v-show="visibleStore" :class="AddRemoveClasses" :play-state="playStateStore" :src="getSrc" :loop="loop" :autoplay="autoplay" :speed="speed" :start-frame="getStartFrame" :end-frame="getEndFrame"></div>
</template>

<script lang="ts">
import { Component, Prop } from "vue-property-decorator";
import lottie from "lottie-web";
import Widget from "./Widget";
import WidgetPlayable from "../mixins/WidgetPlayable";
import { state } from "../mixins/WidgetPlayable";
import { logMessage, logWarning } from "@/helpers/debugHelpers";
import { runWidgetEventConditions } from "../helpers/widgetHelpers";

@Component({ mixins: [WidgetPlayable] })
export default class WidgetAnimatedSvg extends Widget {
	@Prop({ default: "" })
	src: string;

	@Prop({ default: true })
	loop: boolean;

	@Prop({ default: true })
	autoplay: boolean;

	@Prop({ default: 1 })
	speed: number;

	@Prop({ default: 0 })
	startFrame: number;

	@Prop()
	endFrame: number;

	animSrc: string;

	anim: any;

	mounted() {
		this.prevStartFrame = this.propStoreOverride("startFrame") as number;
		this.loadAnimation(this.propStoreOverride("src") as string);
	}

	DoPlayState(s: state) {
		// logMessage("DoPlayState() has been overridden");
		this.play();
	}

	DoStopState(s: state) {
		// logMessage("DoStopState() has been overridden");
		this.stop();
	}

	DoPauseState(s: state) {
		// logMessage("DoPauseState() has been overridden");
		this.pause();
	}

	stop() {
		if (this.anim != undefined) {
			//if playState is set to "stop" in the tag, this can happen
			this.anim.stop();
		} else logWarning(this.id, " - Could not stop(), this.anim is undefined.");
	}

	play() {
		if (this.anim != undefined) {
			//if playState is set to "play" in the tag, this can happen
			const start = this.propStoreOverride("startFrame") as number;
			const end = this.propStoreOverride("endFrame") as number;
			// logMessage(
			//   this.id,
			//   " - WidgetAnimatedSvg: startFrame= ",
			//   start,
			//   ", endFrame= ",
			//   end,
			// );
			//start frame should be 0 if not set. just start playing from the startFrame
			let last_frame = end;
			if (end == undefined) {
				last_frame = this.anim.getDuration(true);
			} else {
				//both startFrame and endFrame are defined, so play that range.
			}
			const segments = [start, last_frame];
			this.anim.playSegments(segments, true);
		} else logWarning(this.id, " - Could not play(), this.anim is undefined.");
	}

	pause() {
		if (this.anim != undefined) {
			//if playState is set to "pause" in the tag, this can happen
			this.anim.pause();
		} else logWarning(this.id, " - Could not pause(), this.anim is undefined.");
	}

	get getSrc(): string {
		const sSrc = this.propStoreOverride("src") as string;
		//logMessage("WidgetAnimatedSvg.get src. return: ", sSrc);
		if (this.$el != undefined) {
			if (sSrc != this.animSrc) {
				logMessage("WidgetAnimatedSvg.get src. loadAnimation: ", sSrc);
				this.loadAnimation(sSrc);
			}
			/*else {
        logWarning(
          "src: ",
          sSrc,
          " is already loaded. Ignoring load request.",
        );
      }*/
		} else {
			logMessage(
				"src: ",
				sSrc,
				", widget has no element, Ignoring load request. This is normal when the widget loads for the first time."
			);
		}
		return sSrc;
	}

	prevStartFrame: number;

	get getStartFrame() {
		// logMessage("----- getStartFrame");
		const sFrame = this.propStoreOverride("startFrame") as number;
		if (this.anim != undefined) {
			const pState = this.propStoreOverride("playState") as string;
			if (pState == "stop" || pState == "default") {
				//animation is stopped, waiting to play, or has finished playing
				if (this.prevStartFrame != sFrame) {
					//anim can be null when the widget starts
					// logMessage(
					//   this.id,
					//   " is going to goToAndStop at ",
					//   sFrame,
					//   ", anim= ",
					//   this.anim,
					// );
					this.anim.goToAndStop(sFrame, true);
					this.anim.renderer.renderFrame(sFrame);
					this.prevStartFrame = sFrame;
				}
			}
			// }
		}
		return sFrame;
	}

	get getEndFrame(): number {
		const eFrame = this.propStoreOverride("endFrame") as number;
		return eFrame;
	}

	loadAnimation(fileSrc: string) {
		if (this.anim) {
			this.anim.destroy();
			this.anim = null;
		}

		logMessage(`loadAnimation:${fileSrc}`, this.id);
		this.anim = lottie.loadAnimation({
			container: this.$el,
			renderer: "svg",
			loop: this.loop,
			autoplay: false, //wait until data_ready before determining when to play
			path: fileSrc,
		});
		this.animSrc = fileSrc;
		this.anim.setSpeed(this.speed);
		this.$emit("animCreated", this.anim);
		this.anim.addEventListener("complete", this.OnAnimationComplete);
		this.anim.addEventListener("loopComplete", this.OnLoopComplete);
		this.anim.addEventListener("data_ready", this.OnDataReady);
	}

	OnAnimationComplete() {
		this.updateWidget("playState", "default"); //if set to "stop" then stopEffects is called.
		runWidgetEventConditions(this, this.$props.endEffects, this.$props.id);
	}

	OnLoopComplete() {
		runWidgetEventConditions(this, this.$props.loopEffects, this.$props.id);
	}

	OnDataReady() {
		//need to incorporate start and end frame. Manually call play() if autoPlay is true.
		// logMessage(this.id, " WidgetAnimatedSvg data is ready");
		// logMessage(this.id, " WidgetAnimatedSvg autoPlay is ", this.autoplay);
		const pState = this.propStoreOverride("playState") as string;
		if (this.autoplay || pState == "play") this.play();
		else {
			//autoplay is false, so go to the start frame and stop
			const sFrame = this.propStoreOverride("startFrame") as number;
			this.anim.goToAndStop(sFrame, true);
		}
	}

	beforeDestroy() {
		logMessage("destroying animatedSVG", this.id);
		if (this.anim) this.anim.destroy();
		// Destroying lottie was causing a problem on experiment pages in Photosynthesis with SVGs and links to glossary pages with SVGs.
		// When the second glossary page was opened, the svg on the first glossary page would destoy lottie,
		// which would kill the SVGs on the case page.
		// So it's probably not a good idea to destroy lottie.
		// if (lottie) lottie.destroy();
	}
}
</script>

<style>
/* in safari we need to apply this style to the child <svg> element */
.imageOutline > svg {
	filter: drop-shadow(0.2rem 0.2rem 0 var(--color-blue4))
		drop-shadow(-0.2rem 0.2rem 0 var(--color-blue4))
		drop-shadow(0.2rem -0.2rem 0 var(--color-blue4))
		drop-shadow(-0.2rem -0.2rem 0 var(--color-blue4));
	-webkit-filter: drop-shadow(0.2rem 0.2rem 0 var(--color-blue4))
		drop-shadow(-0.2rem 0.2rem 0 var(--color-blue4))
		drop-shadow(0.2rem -0.2rem 0 var(--color-blue4))
		drop-shadow(-0.2rem -0.2rem 0 var(--color-blue4));
}
</style>
