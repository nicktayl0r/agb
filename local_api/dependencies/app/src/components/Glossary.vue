<template>
	<div id="glossaryParent" v-show="show">
		<div id="glossaryBG" @click="toggleShow"></div>
		<div id="glossaryBody">
			<div id="glossaryTitleBar">
				<b>Glossary</b>
				<widget-button id="glossaryCloseButton" class="buttonX white" @click.native="toggleShow" />
			</div>
			<div id="glossaryContent">
				<div id="glossaryDefinitionTitle">
					<b v-html="currentTermName"></b>
				</div>
				<div id="glossaryDefinition">
					<glossary-page-async />
				</div>
				<div id="glossaryRelatedTerms">
					<span v-if="RelatedTerms.length > 0">Related terms: </span>
					<span v-for="(relatedTerm, index) in RelatedTerms" :key="relatedTerm.pageID">
						<router-link :id="relatedTerm.pageID" :class="{'internalLink':true, glossarySelectedTerm: relatedTerm.pageID === currentTermID}" @click.native="clickedRelatedTerm(relatedTerm.pageID)" to="" :title='relatedTerm.name'>
							<slot /><span v-html="relatedTerm.name"></span>
						</router-link>
						<span v-if="index + 1 < RelatedTerms.length">, </span>

					</span>
				</div>
				<div id="glossarySidebarSearch">
					<input id="searchInput" ref="searchInput" type="text" placeholder="Search" v-model.trim="searchQuery">
					<widget-button id="glossarySearchClearButton" class="buttonX glossarySearchClear" @click.native="clearSearch" />
				</div>
				<div id="glossarySidebar">
					<div id="termsScroll" ref="termsScroll">
						<div v-for="term in Terms" :key="term.pageID" class="glossaryTermLink">
							<router-link :id="term.pageID" :class="{'internalLink':true, glossarySelectedTerm: term.pageID === currentTermID}" @click.native="clickedRelatedTerm(term.pageID)" to="" :title='term.name'>
								<slot /><span v-html="term.name"></span>
							</router-link>
						</div>
					</div>
				</div>
			</div>
		</div>
	</div>
</template>

<script lang="ts">
import { Component, Prop } from "vue-property-decorator";
import Vue from "vue";
import { Maybe } from "true-myth";

import GlossaryPageAsync from "@/components/GlossaryPageAsync.vue";
import WidgetButton from "@/components/WidgetButton.vue";
import WidgetLink from "@/components/WidgetLink.vue";
import {
	GlossaryModule,
	GlossaryTrack,
	GlossaryPage,
	GlossaryRelatedTerm,
} from "@/data models/projectModels";
import {
	getCurrentGlossaryTrack,
	getCurrentGlossaryPage,
} from "@/helpers/userHelpers";
import { readGlossary } from "@/store/modules/appData";
import {
	readModulesState,
	commitGlossaryOpenState,
} from "@/store/modules/userData";
import { commitSetModuleCurrentPage } from "@/store/modules/userData";

@Component({
	components: {
		GlossaryPageAsync,
		WidgetButton,
		WidgetLink,
	},
})
export default class Glossary extends Vue {
	searchQuery = "";

	show = false;

	toggleShow(): void {
		this.show = !this.show;
		commitGlossaryOpenState(this.$store, this.show);
	}

	showHide(val: boolean): void {
		this.show = val;
		commitGlossaryOpenState(this.$store, this.show);

		if (this.show) {
			this.$nextTick(() => {
				const termEl = document.getElementById(this.currentTermID);
				if (termEl) {
					// center the term in the parent scroll container
					const termPos = termEl.offsetTop;
					const termHalfHeight = termEl.getBoundingClientRect().height * 0.5;
					const scrollParent = this.$refs.termsScroll as Element;
					const parentHalfHeight = scrollParent.clientHeight * 0.5;
					const scrollPos = termPos + termHalfHeight - parentHalfHeight;
					scrollParent.scrollTop = scrollPos;
				}
			});
		}
	}

	hideIfOpen(): void {
		if (this.show) this.showHide(false);
	}

	clearSearch() {
		this.searchQuery = "";
	}

	get Terms(): GlossaryPage[] {
		const currentTrack = getCurrentGlossaryTrack();
		if (currentTrack && currentTrack.pages) {
			if (
				this.searchQuery &&
				this.searchQuery != "" &&
				this.searchQuery != " "
			) {
				try {
					const includesQueryRegEx = new RegExp(this.searchQuery, "i"); // case insensitive
					return currentTrack.pages.filter((page) =>
						includesQueryRegEx.test(page.name)
					);
				} catch (ex) {
					return [];
				}
			}
			return currentTrack.pages;
		}
		return [];
	}

	get RelatedTerms(): GlossaryRelatedTerm[] {
		const currentGlossaryPage = getCurrentGlossaryPage();
		if (currentGlossaryPage && currentGlossaryPage.relatedTerms)
			return currentGlossaryPage.relatedTerms;
		return [];
	}

	get currentTermID(): string {
		const currentGlossaryPage = getCurrentGlossaryPage();
		if (currentGlossaryPage && currentGlossaryPage.pageID)
			return currentGlossaryPage.pageID;
		return "";
	}

	get currentTermName(): string {
		const currentGlossaryPage = getCurrentGlossaryPage();
		if (currentGlossaryPage && currentGlossaryPage.name)
			return currentGlossaryPage.name;
		return "";
	}

	clickedRelatedTerm(termID: string) {
		if (!this.$store) return;

		commitSetModuleCurrentPage(this.$store, {
			moduleName: "glossary",
			pageID: termID,
		});
	}
}
</script>

<style>
#glossaryParent {
	display: flex;
	justify-content: center;
	align-items: center;
	position: relative;
}
#glossaryBG {
	position: absolute;
	background-color: rgba(0, 0, 0, 0.75);
	z-index: 200;
	width: 100%;
	height: 100%;
}
#glossaryBody {
	background-color: #5c4866;
	color: white;
	width: 90%;
	height: 90%;
	display: flex;
	flex-direction: column;
	position: absolute;
	top: 5%;
	left: 5%;
	z-index: 220;
	border-radius: 0.626rem;
}
#glossaryTitleBar {
	padding: 1.137vh;
}
#glossaryContent {
	background-color: #74617a;
	margin: 0 1.137vh 1.137vh 1.137vh;
	padding: 2.25vh;
	display: grid;
	grid-template-columns: auto 22.75vh;
	grid-template-rows: min-content 1fr min-content;
	flex-grow: 1;
	max-height: 100%;
	min-height: 0;
	height: 100%;
	border-radius: 0.626rem;
}
#glossaryDefinitionTitle {
	grid-column: 1 / span 1;
	grid-row: 1 / span 1;
}
#glossaryDefinition {
	grid-column: 1 / span 1;
	grid-row: 2 / span 1;
	max-height: 100%;
	overflow-y: auto;
	padding-right: 2.25vh;
}
#glossaryRelatedTerms {
	grid-column: 1 / span 1;
	grid-row: 3 / span 1;
	align-self: center;
	padding-top: 2.25vh;
}
#glossarySidebarSearch {
	grid-column: 2 / span 1;
	grid-row: 1 / span 1;
	padding-left: 2.25vh;
	padding-bottom: 0.2rem;
	border-left: 1px solid #5c4866;
	position: relative;
}
#glossarySidebar {
	grid-column: 2 / span 1;
	grid-row: 2 / span 2;
	border-left: 1px solid #5c4866;
	padding-left: 2.25vh;
	min-height: 0;
	overflow: hidden;
}
#termsScroll {
	position: relative;
	max-height: 100%;
	overflow-y: auto;
	overflow-x: hidden;
}
#glossaryCloseButton {
	position: absolute;
	right: 0;
	top: 0;
}
#searchInput {
  padding-right: 3.3vh;
}
.glossarySelectedTerm {
	color: #5c4866 !important;
	background-color: white;
	border-radius: 0.15em;
	padding: 0.25em;
	box-decoration-break: clone;
	-webkit-box-decoration-break: clone;
}
.glossarySelectedTerm span {
	position: relative;
}
.glossaryTermLink {
	margin: 0.75em 0 0.75em 0;
}
.glossarySearchClear {
	position: absolute;
	top: 0.57vh;
	right: 0.57vh;
}
.glossarySearchClear .bgImage {
	width: 0.75rem !important;
	height: 1.5rem !important;
}
</style>
