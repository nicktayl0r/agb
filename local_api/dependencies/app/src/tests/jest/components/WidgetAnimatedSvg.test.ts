import { getLocalVue, getCleanStore } from "@/tests/jest/jestSetup";
const localVue = getLocalVue();
const store = getCleanStore();

import "jest-canvas-mock";

import { shallowMount } from "@vue/test-utils";
import { WidgetBaseTests } from "@/tests/jest/components/WidgetBaseTests";
import { MixinWidgetPlayableTests } from "@/tests/jest/mixins/MixinWidgetPlayableTests";

import WidgetAnimatedSvg from "@/components/WidgetAnimatedSvg.vue";

describe("WidgetAnimatedSvg", () => {
  WidgetBaseTests(WidgetAnimatedSvg);

  it("matches html snapshot", async () => {
    const wrapper = shallowMount(WidgetAnimatedSvg, {
      propsData: {
        id: "testWidget",
        src: "././dist/assets/svg/thermometer_AnimatedSVG.json"
      },
      localVue,
      store
    }); /*?.*/
    expect(wrapper.element).toMatchSnapshot(); /*?.*/
  });

  MixinWidgetPlayableTests(WidgetAnimatedSvg);
});
