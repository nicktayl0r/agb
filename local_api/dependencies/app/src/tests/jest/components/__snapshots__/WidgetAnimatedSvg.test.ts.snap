// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetAnimatedSvg matches html snapshot 1`] = `
<div
  autoplay="autoplay"
  class=""
  id="testWidget"
  loop="loop"
  play-state="default"
  speed="1"
  src="././dist/assets/svg/thermometer_AnimatedSVG.json"
  start-frame="0"
/>
`;
